File: DoryApi/api/espansioni.py
from flask import Blueprint, current_app, request, jsonify
from flask_jwt_extended import jwt_required
import mysql
from db import get_db_connection


espansioni_bp = Blueprint('espansioni', __name__, url_prefix='/api/espansioni')

# Recupera tutte le espansioni disponibili
@espansioni_bp.route('', methods=['GET'])
@jwt_required()
def get_espansioni():
    conn = get_db_connection()
    cur = conn.cursor()
    cur.execute("SELECT EXP_COD, EXP_NAME FROM espansioni ORDER BY EXP_NAME")
    results = [{"EXP_COD": r[0], "EXP_NAME": r[1]} for r in cur.fetchall()]
    cur.close()
    conn.close()
    return jsonify(results)


# Recupera le espansioni associate a un pannello
@espansioni_bp.route('/pannello/<int:pan_cod>', methods=['GET'])
@jwt_required()
def get_espansioni_pannello(pan_cod):
    conn = get_db_connection()
    cur = conn.cursor()
    query = """
        SELECT ep.EPA_EXP, e.EXP_NAME, ep.EPA_ADD
        FROM espansioni_pannelli ep
        JOIN espansioni e ON e.EXP_COD = ep.EPA_EXP
        WHERE ep.EPA_PAN = %s
        ORDER BY e.EXP_NAME
    """
    cur.execute(query, (pan_cod,))
    results = [{"EXP_COD": r[0], "EXP_NAME": r[1], "i2cAddr": r[2]} for r in cur.fetchall()]
    cur.close()
    conn.close()
    return jsonify(results)


# Aggiunge un'espansione a un pannello
@espansioni_bp.route('/pannelli/<int:pan_cod>', methods=['POST'])
@jwt_required()
def add_espansione_pannello(pan_cod):
    data = request.get_json()
    exp_cod = data.get('exp_cod')
    i2c = data.get('i2cAddr')

    if not all([exp_cod, i2c]):
        return jsonify({'msg': 'Dati mancanti'}), 400

    conn = get_db_connection()
    cur = conn.cursor()
    try:
        cur.execute(
            "INSERT INTO espansioni_pannelli (EPA_PAN, EPA_EXP, EPA_ADD) VALUES (%s,%s,%s)",
            (pan_cod, exp_cod, i2c)
        )
        conn.commit()
    except Exception as e:
        cur.close()
        conn.close()
        return jsonify({'msg': str(e)}), 400

    cur.close()
    conn.close()
    return jsonify({'status': 'ok'})


# Rimuove un'espansione da un pannello
@espansioni_bp.route('/pannelli/<int:pan_cod>', methods=['DELETE'])
@jwt_required()
def remove_espansione_pannello(pan_cod):
    data = request.get_json()
    exp_cod = data.get('exp_cod')
    i2c = data.get('i2cAddr')

    if not all([exp_cod, i2c]):
        return jsonify({'msg': 'Dati mancanti'}), 400

    conn = get_db_connection()
    cur = conn.cursor()
    try:
        cur.execute(
            "DELETE FROM espansioni_pannelli WHERE EPA_PAN=%s AND EPA_EXP=%s AND EPA_ADD=%s",
            (pan_cod, exp_cod, i2c)
        )
        conn.commit()
    except Exception as e:
        cur.close()
        conn.close()
        return jsonify({'msg': str(e)}), 400

    cur.close()
    conn.close()
    return jsonify({'status': 'ok'})

File: DoryGestionale/src/api/espansioni.js
import apiClient from './auth'
const API_BASE = 'http://localhost:5000/api';

// GET /api/espansioni
// - Richiede: Solo token JWT
// - Risponde: [{ EXP_COD, EXP_NAME }, ...]
// - NOTA: Recupera tutte le espansioni disponibili
export const getEspansioni = async () => {
  const token = localStorage.getItem('token');
  const res = await apiClient.get(`${API_BASE}/espansioni`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return res.data; // array di oggetti {EXP_COD, EXP_NAME}
};

// GET /api/espansioni/pannello/<pan_cod>
// - Richiede: Solo token JWT
// - Risponde: [{ EXP_COD, EXP_NAME, i2cAddr }, ...]
// - NOTA: Recupera le espansioni già associate a un pannello
export const getEspansioniPannello = async (pan_cod) => {
  const token = localStorage.getItem('token');
  const res = await apiClient.get(`${API_BASE}/espansioni/pannello/${pan_cod}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  return res.data;
};


// POST /api/espansioni/pannelli/<pan_cod>
// - Richiede: { exp_cod, i2cAddr }
// - Risponde: espansione aggiunta
// - NOTA: Associa un'espansione ad un pannello
export const addEspansionePannello = async (pan_cod, exp_cod, i2cAddr) => {
  const token = localStorage.getItem('token');
  const res = await apiClient.post(`${API_BASE}/espansioni/pannelli/${pan_cod}`, { exp_cod, i2cAddr }, {
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`,
    }
  });
  return res.data;
};


// DELETE /api/espansioni/pannelli/<pan_cod>
// - Richiede: { exp_cod, i2cAddr }
// - Risponde: espansione rimossa
// - NOTA: Rimuove l'associazione di un'espansione da un pannello
export const removeEspansionePannello = async (pan_cod, exp_cod, i2cAddr) => {
  const token = localStorage.getItem('token');
  const res = await apiClient.delete(`${API_BASE}/espansioni/pannelli/${pan_cod}`, {
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`,
    },
    data: { exp_cod, i2cAddr }
  });
  return res.data;
};